### PR TITLE
[feat] refactored extension module

### PR DESCRIPTION
**Pull request by FrankLeeeee**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: extensions/cpp_extension.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        """
+
+        # get the current file path
+        # iteratively check the parent directory
+        # if the parent directory is "extensions", then the current file path is the root directory
+        # otherwise, the current file path is inside the root directory
+        current_file_path = Path(__file__)
+        while True:
+            if current_file_path.name == "extensions":
+                current_file_path.parent
```
Comment (ver217) [inline extensions/cpp_extension.py:40]: What's the effect of this line

File: extensions/cpp_extension.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    def cxx_flags(self) -> List[str]:
+        """
+        This function should return a list of cxx compilation flags for extensions.
+        """
+
+    @abstractmethod
+    def nvcc_flags(self) -> List[str]:
+        """
+        This function should return a list of nvcc compilation flags for extensions.
+        """
```
Comment (ver217) [inline extensions/cpp_extension.py:132]: cpp extensions should not have nvcc flags, right?

File: extensions/cpu_adam/cpu_adam_x86.py [hunk 3]
```diff
@@ -1,38 +1,27 @@
...
-    def build(self):
-        self.kernel_builder.build()
-        self._requires_build = True
+from ..cuda_extension import _CudaExtension
+from ..utils import append_nvcc_threads
 
-    def load(self):
-        return self.kernel_builder.load()
 
+class CpuAdamX86Extension(_CudaExtension):
```
Comment (ver217) [inline extensions/cpu_adam/cpu_adam_x86.py:7]: Why does cpu adam inherit cuda extention?

File: extensions/cuda_extension.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            cuda_available = torch.cuda.is_available()
+        except:
+            cuda_available = False
+        return cuda_available
+
+    def assert_hardware_compatible(self) -> None:
+        from torch.utils.cpp_extension import CUDA_HOME
+
+        if not CUDA_HOME:
+            raise RuntimeError(
```
Comment (ver217) [inline extensions/cuda_extension.py:28]: As this method is `assert_hardware_compatible`, should we raise assertation error here?

File: op_builder/builder.py [hunk 5]
```diff
@@ -218,12 +218,7 @@ def builder(self) -> Union["CUDAExtension", "CppExtension"]:
         from torch.utils.cpp_extension import CppExtension, CUDAExtension
 
         if self.ext_type == "cpp":
-            return CppExtension(
```
Comment (ver217) [inline op_builder/builder.py:221]: It seems this "if branch" is empty

File: extensions/cpp_extension.py (hunk 1 above)
Comment (FrankLeeeee) [reply extensions/cpp_extension.py]: This line is removed.

File: extensions/cpp_extension.py (hunk 2 above)
Comment (FrankLeeeee) [reply extensions/cpp_extension.py]: `nvcc_flags` is moved to cuda ext class.

File: extensions/cuda_extension.py (hunk 4 above)
Comment (FrankLeeeee) [reply extensions/cuda_extension.py]: changed to assertion error.

File: op_builder/builder.py (hunk 5 above)
Comment (FrankLeeeee) [reply op_builder/builder.py:221]: `op_builder` is removed, everything is collated into `extensions`.

File: extensions/cpu_adam/cpu_adam_x86.py (hunk 3 above)
Comment (FrankLeeeee) [reply extensions/cpu_adam/cpu_adam_x86.py:7]: Cuz CPU adam needs to include cuda header files.

Screenshot 2024-01-24 at 15 17 18

File: op_builder/builder.py (hunk 5 above)
Comment (ver217) [reply op_builder/builder.py:221]: can we remove this file?

Comment (FrankLeeeee) [reply op_builder/builder.py:221]: done

File: extensions/cpp_extension.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            print(f"[extension] Compiling the JIT {self.name} kernel during runtime now")
+
+        build_start = time.time()
+        op_kernel = load(
+            name=self.name,
+            sources=self.strip_empty_entries(self.sources_files()),
+            extra_include_paths=self.strip_empty_entries(self.include_dirs()),
+            extra_cflags=self.cxx_flags(),
+            extra_cuda_cflags=self.nvcc_flags(),
```
Comment (ver217) [inline extensions/cpp_extension.py:95]: Will this cause error?

**Comment by FrankLeeeee** [comment]
Screenshot 2024-01-25 at 15 24 55
Screenshot 2024-01-25 at 15 28 24
Screenshot 2024-01-25 at 15 30 11
Screenshot 2024-01-25 at 15 31 00
 
Screenshot 2024-01-25 at 15 31 56
Screenshot 2024-01-25 at 15 32 51
Screenshot 2024-01-25 at 15 40 26
Screenshot 2024-01-25 at 15 41 44
Screenshot 2024-01-25 at 15 44 24
Screenshot 2024-01-25 at 16 07 03
Screenshot 2024-01-25 at 16 09 00
Screenshot 2024-01-25 at 16 09 18
Screenshot 2024-01-25 at 16 16 55


2 tests failed due to precision on the H800 system.

File: extensions/cpp_extension.py (hunk 6 above)
Comment (FrankLeeeee) [reply extensions/cpp_extension.py:95]: Why would it cause error?

**Review by ver217 — APPROVED** [review]

**Pull request merged**